### PR TITLE
[feature] add gptq for inference

### PR DESCRIPTION
**Pull request by Xu-Kai**
## 📌 Checklist before creating the PR

- [ ] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [ ] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

Closes #4416,   closes #4494,  closes #4538,  closes #4695, closes #4706, closes #4716

## 📝 What does this PR do?

- add gptq for inference 
- combine gptq and kv cache

## 💥 Checklist before requesting a review

- [ ] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [ ] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [ ] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: colossalai/inference/quant/gptq/cai_gptq/gptq_triton.py [hunk 1]
```diff
@@ -0,0 +1,558 @@
+import torch
+import triton
+import triton.language as tl
+from auto_gptq.nn_modules.triton_utils import custom_autotune
+
+# from ..ops.triton.kernels.activations_kernels import relu,  gelu, silu
+# code based https://github.com/fpgaminer/GPTQ-triton
```
Comment (CjhHa1) [inline colossalai/inference/quant/gptq/cai_gptq/gptq_triton.py:7]: remove this?

Comment (Xu-Kai) [reply colossalai/inference/quant/gptq/cai_gptq/gptq_triton.py]: Done.

File: colossalai/kernel/cuda_native/csrc/gptq/cu_compat.cuh [hunk 2]
```diff
@@ -0,0 +1,58 @@
+// Adapted from turboderp exllama: https://github.com/turboderp/exllama
```
Comment (ver217) [inline colossalai/kernel/cuda_native/csrc/gptq/cu_compat.cuh:1]: I notice you use code from exllama. Do you need to add their license to ours?

Comment (Xu-Kai) [reply colossalai/kernel/cuda_native/csrc/gptq/cu_compat.cuh:1]: I added their license in our LICENSE.

**Review by ver217 — APPROVED** [review]

**Review by tiandiao123 — APPROVED** [review]

**Pull request merged**